src/stream_connecter_base.cpp

```diff
@@ -0,0 +1,167 @@
+*/
+
+#include "precompiled.hpp"
+#include "stream_connecter_base.hpp"
+#include "session_base.hpp"
+#include "address.hpp"
+#include "random.hpp"
+
+zmq::stream_connecter_base_t::stream_connecter_base_t (
+  zmq::io_thread_t *io_thread_,
+  zmq::session_base_t *session_,
+  const zmq::options_t &options_,
+  zmq::address_t *addr_,
+  bool delayed_start_) :
+    own_t (io_thread_, options_),
+    io_object_t (io_thread_),
+    _addr (addr_),
+    _s (retired_fd),
+    _handle (static_cast<handle_t> (NULL)),
+    _socket (session_->get_socket ()),
+    _delayed_start (delayed_start_),
+    _reconnect_timer_started (false),
+    _session (session_),
+    _current_reconnect_ivl (options.reconnect_ivl)
+{
+    zmq_assert (_addr);
+    _addr->to_string (_endpoint);
+    // TODO the return value is unused! what if it fails? if this is impossible
+    // or does not matter, change such that endpoint in initialized using an
+    // initializer, and make endpoint const
+}
+
+zmq::stream_connecter_base_t::~stream_connecter_base_t ()
+{
+    zmq_assert (!_reconnect_timer_started);
+    zmq_assert (!_handle);
+    zmq_assert (_s == retired_fd);
+}
+
+void zmq::stream_connecter_base_t::process_plug ()
+{
+    if (_delayed_start)
+        add_reconnect_timer ();
+    else
+        start_connecting ();
+}
+
+void zmq::stream_connecter_base_t::process_term (int linger_)
+{
+    if (_reconnect_timer_started) {
+        cancel_timer (reconnect_timer_id);
+        _reconnect_timer_started = false;
+    }
+
+    if (_handle) {
+        rm_handle ();
+    }
+
+    if (_s != retired_fd)
+        close ();
+
+    own_t::process_term (linger_);
+}
+
+void zmq::stream_connecter_base_t::add_reconnect_timer ()
+{
+    if (options.reconnect_ivl != -1) {
+        const int interval = get_new_reconnect_ivl ();
+        add_timer (interval, reconnect_timer_id);
+        _socket->event_connect_retried (_endpoint, interval);
+        _reconnect_timer_started = true;
+    }
+}
+
+int zmq::stream_connecter_base_t::get_new_reconnect_ivl ()
+{
+    //  The new interval is the current interval + random value.
+    const int interval =
+      _current_reconnect_ivl + generate_random () % options.reconnect_ivl;
+
+    //  Only change the current reconnect interval  if the maximum reconnect
+    //  interval was set and if it's larger than the reconnect interval.
+    if (options.reconnect_ivl_max > 0
+        && options.reconnect_ivl_max > options.reconnect_ivl)
+        //  Calculate the next interval
+        _current_reconnect_ivl =
+          std::min (_current_reconnect_ivl * 2, options.reconnect_ivl_max);
+    return interval;
+}
+
+void zmq::stream_connecter_base_t::rm_handle ()
+{
+    rm_fd (_handle);
+    _handle = static_cast<handle_t> (NULL);
+}
+
+void zmq::stream_connecter_base_t::close ()
+{
+    zmq_assert (_s != retired_fd);
+#ifdef ZMQ_HAVE_WINDOWS
+    const int rc = closesocket (_s);
+    wsa_assert (rc != SOCKET_ERROR);
+#else
+    const int rc = ::close (_s);
```

Comment:
@sigiesec this breaks the build in any recent distro:

```
[  233s] src/stream_connecter_base.cpp: In member function 'void zmq::stream_connecter_base_t::close()':
[  233s] src/stream_connecter_base.cpp:132:22: error: '::close' has not been declared
[  233s]      const int rc = ::close (_s);
[  233s]                       ^~~~~
[  233s] src/stream_connecter_base.cpp:132:22: note: suggested alternative: 'clone'
[  233s]      const int rc = ::close (_s);
[  233s]                       ^~~~~
[  233s]                       clone
```

https://build.opensuse.org/package/live_build_log/network:messaging:zeromq:git-stable/libzmq/Debian_Next/x86_64

Reply:
#include <unistd.h> is probably missing there

Reply:
#3388